### PR TITLE
Image moderation metrics

### PR DESCRIPTION
**Pull request by islemaster**
Logs Azure content moderation activity to Redshift.  This serves a few purposes:

- Monitoring how much content is being blocked on our site by automated moderation.
- Comparing our own metrics to Azure's in case of any billing disputes.
- Checking the distribution of confidence scores to help us fine-tune our filter if needed.
- Tracking request duration so we'll be able to tell if Azure's service is responsible for slow requests.

## Details

- **study:** `azure-content-moderation`
- **study_group:** `v1` - To bump anytime we make changes that affect these metrics.

I'm logging two events:
```
event: 'moderation-request
```
This happens right before we send a request to Azure's service.  I'd expect the event count to be equal to (or maybe slightly higher than) the request counts reported by Azure for billing purposes.

```
event: 'moderation-result',
data_string: 'everyone' | 'racy' | 'adult'
data_json: {
  "AdultClassificationScore": <number> (0.0-1.0),
  "IsImageAdultClassified": <boolean>,
  "RacyClassificationScore": <number> (0.0-1.0),
  "IsImageRacyClassified": <boolean>,
  "RequestDuration": <number> (seconds)
}
```
This happens whenever we get a rating back from Azure's service.  This count will be lower than the request count; the difference represents our error rate.  The `data_string` represents the final rating given to the image by our system; right now this maps exactly to the boolean values returned by Azure, but in the future we might fine-tune it based on the numeric scores or other factors.  The values in the JSON are the raw values sent back by Azure's API.

## Related work
* Feature introduced in https://github.com/code-dot-org/code-dot-org/pull/21812
* Fix unhandled exception in https://github.com/code-dot-org/code-dot-org/pull/21884
* Tests moved to their own file in https://github.com/code-dot-org/code-dot-org/pull/21894
* Moderation limited to particular app types in https://github.com/code-dot-org/code-dot-org/pull/21900
* Request pattern optimized in https://github.com/code-dot-org/code-dot-org/pull/21914

### CODE REVIEW

**Comment by islemaster** [comment]
@bencodeorg Please ignore the code diff for now.  Based on the PR description, is there anything else I should be logging when tracking our interactions with this external API?

File: lib/cdo/azure_content_moderator.rb [hunk 1]
```diff
@@ -32,8 +38,30 @@ def rate_image(image_data, content_type)
...
-    uri = URI(@endpoint + '/moderate/v1.0/ProcessImage/Evaluate')
+    report_request
+
+    request_start_time = Time.now
+    result = make_request(image_data, content_type)
+    request_duration = Time.now - request_start_time
+
+    rating = rating_from_azure_result(result)
+    report_response(rating, result, request_duration)
+    rating
```
Comment (davidsbailey) [inline lib/cdo/azure_content_moderator.rb]: nice self-documenting method names 👍

File: lib/test/cdo/test_azure_content_moderator.rb [hunk 2]
```diff
@@ -43,9 +48,40 @@ def test_raise_on_image_too_small
...
+      data[:study] == 'azure-content-moderation' &&
+        data[:study_group] == 'v1' &&
+        data[:event] == 'moderation-result' &&
+        %w(everyone racy adult).include?(data[:data_string]) &&
+        JSON.parse(data[:data_json]).keys == %w(
+          AdultClassificationScore
+          IsImageAdultClassified
+          RacyClassificationScore
+          IsImageRacyClassified
+          RequestDuration
```
Comment (davidsbailey) [inline lib/test/cdo/test_azure_content_moderator.rb:92]: this seems like a good level of detail to enforce.

**Comment by islemaster** [comment]
Yep!  The idea is we can derive total error count as (requests - results) and error details end up in Honeybadger, which is better for tracking down issues anyway.

**Comment by bencodeorg** [comment]
Based on the questions you listed in the description, I think the data you're logging looks good for answering them. 

Is it possible (and if possible, useful?) to log some identifier that would point us to the actual content that was moderated (like a S3 URL or something?) Also there's no ability to link between a request and a response as instrumented here -- would it be helpful to know what type(s) of requests are erroring out?

**Comment by islemaster** [comment]
I can add an image URL to these events, which may be useful for the "fine-tune our filter" purpose in particular.  I imagine at some point we might pull a report of 'borderline' content to decide whether we need to loosen or tighten our controls a bit.  I assume it's fine for me to throw that into the JSON?

That will also allow us to pair request/result events, although I'm not sure it will be useful.  Errors will go to Honeybadger which will implicitly provide us with request URLs, backtrace, and other context specifically useful to debugging.

**Comment by bencodeorg** [comment]
Yup that sounds like a good plan to me!

**Comment by davidsbailey** [comment]
Great questions Ben! Brad, I agree, although my initial intuition was also to somehow link request events to response events, I don't believe there is a need.

**Pull request merged**